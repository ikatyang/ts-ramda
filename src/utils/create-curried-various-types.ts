import * as dts from 'dts-element';
import {isEqual} from 'lodash';
import {create_curried_types} from './create-curried-types';
import {create_masks} from './create-masks';
import {create_selectable_signatures} from './create-selectable-signatures';

const is_object_type = (element: any): element is dts.IObjectType =>
  (element.kind === dts.ElementKind.ObjectType);
const is_object_member = (element: any): element is dts.IObjectMember =>
  (element.kind === dts.ElementKind.ObjectMember);

export const create_various_curried_types = (name: string, types: {[kind: string]: dts.IFunctionType}) => {
  const keys = Object.keys(types);

  const parameters_length = keys.reduce<number>(
    (length, key) =>
      (length === -1 || length === types[key].parameters!.length)
        ? types[key].parameters!.length
        : (() => {
          throw new Error(`curried function types must be the same length`);
        })(),
    -1,
  );

  const curried_types_declarations = keys.reduce<dts.ITypeDeclaration[][]>(
    (reduced, key) => [...reduced, create_curried_types(`${name}_${key}`, types[key], false)],
    [],
  );

  const masks = create_masks(parameters_length);
  const reverse_masks = masks.reduce<{[mask: string]: number}>(
    (current_reverse_masks, mask, index) => ({...current_reverse_masks, [mask]: index}),
    {},
  );

  const types_parameters = keys.reduce<dts.IParameterDeclaration[][]>(
    (reduced, key) => [...reduced, types[key].parameters!],
    [],
  );

  const non_conflicts = [...new Array(parameters_length)].map(
    (_, index) => {
      const parameter = types_parameters[0][index];
      return types_parameters.slice(1).every(type_parameters => isEqual(type_parameters[index], parameter));
    },
  );

  const conflict_declarations: dts.ITypeDeclaration[][] = [];
  const conflict_declarations_selectables: dts.IObjectMember[][][] = [];

  const non_conflict_rebases: {[name: string]: string} = {};
  const non_conflict_declarations: dts.ITypeDeclaration[] = [];
  const non_conflict_declarations_selectables: dts.IObjectMember[][] = [];

  curried_types_declarations[0].forEach((curried_type_declaration, index) => {
    const mask = masks[index];
    if (mask.split('').every((bit, bit_index) => bit === '0' || non_conflicts[bit_index])) {
      const rebase_name = `${name}_${mask}`;
      non_conflict_declarations.push(dts.create_type_declaration({
        name: rebase_name,
        generics: curried_type_declaration.generics,
        type: dts.create_object_type({
          members: curried_types_declarations
            .map(curried_types_declaration =>
              (curried_types_declaration[index].type! as dts.IObjectType).members! as dts.IObjectMember[])
            .reduce((a, b) => [...a, ...b], []),
        }),
      }));
      non_conflict_declarations_selectables.push(
        curried_types_declarations
          .map((curried_types_declaration, type_index) =>
            create_selectable_signatures(
              (curried_types_declaration[index].type! as dts.IObjectType).members! as dts.IObjectMember[],
              keys[type_index],
            ),
          )
          .reduce((a, b) => [...a, ...b], []),
      );
      curried_types_declarations.map(x => x[index]).forEach(type_declaration => {
        non_conflict_rebases[type_declaration.name] = rebase_name;
      });
    } else {
      conflict_declarations.push(curried_types_declarations
        .map(curried_type_declarations => curried_type_declarations[index]),
      );
      conflict_declarations_selectables.push(curried_types_declarations
        .map((curried_type_declarations, type_index) => {
          const members = (curried_type_declarations[index].type as dts.IObjectType).members;
          return members !== undefined && members.every(is_object_member)
            ? create_selectable_signatures(
              (members as dts.IObjectMember[]) || [],
            )
            : [];
        }),
      );
    }
  });

  const mixed_type_declarations_selectables = [
    ...non_conflict_declarations_selectables,
    ...conflict_declarations_selectables.reduce((a, b) => [...a, ...b], []),
  ];
  const mixed_type_declarations = [
    ...non_conflict_declarations,
    ...conflict_declarations.reduce((a, b) => [...a, ...b], []),
  ].map((type_declaration, index) => {
    const object_type = type_declaration.type;
    if (!is_object_type(object_type) || object_type.members!.some(member => !is_object_member(member))) {
      return type_declaration;
    }
    const repeated_function_types: {[name: string]: dts.IFunctionType} = {};
    const members: dts.IObjectMember[] = [];
    object_type.members!.forEach((member: dts.IObjectMember, member_index, origin_members) => {
      const function_type = (member.owned as dts.IFunctionDeclaration).type!;
      const function_return_type = function_type.return! as dts.IGeneralType;
      if (function_return_type.name in non_conflict_rebases) {
        repeated_function_types[non_conflict_rebases[function_return_type.name]] = function_type;
      } else {
        members.push(
          member,
          ...(
            (origin_members.length > 1)
              ? [mixed_type_declarations_selectables[index][member_index]]
              : []
          ),
        );
      }
    });
    const non_conflict_merged_members = Object.keys(repeated_function_types).map(key => {
      const function_type = repeated_function_types[key];
      const function_return_type = function_type.return as dts.IGeneralType;
      return dts.create_object_member({
        owned: dts.create_function_declaration({
          name: undefined,
          type: dts.create_function_type({
            ...function_type,
            return: dts.create_general_type({
              name: key,
              generics: function_return_type.generics,
            }),
          }),
        }),
      });
    });
    members.push(
      ...non_conflict_merged_members,
      ...create_selectable_signatures(non_conflict_merged_members),
    );
    object_type.members = members;
    return type_declaration;
  });

  return mixed_type_declarations;
};
