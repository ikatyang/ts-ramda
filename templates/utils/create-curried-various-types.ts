import * as dts from 'dts-element';
import * as R from 'ramda';
import { create_curried_types } from './create-curried-types';
import { create_masks } from './create-masks';
import { create_selectable_signatures } from './create-selectable-signatures';
import { is_object_member, is_object_type } from './is-element';

export const create_various_curried_types = (
  name: string,
  types: { [kind: string]: dts.IFunctionType },
  selectable = true,
  placeholder = true,
) => {
  const keys = Object.keys(types);

  const parameters_length = keys.reduce<number>(
    (length, key) =>
      length === -1 || length === types[key].parameters!.length
        ? types[key].parameters!.length
        : (() => {
            throw new Error(`curried function types must be the same length`);
          })(),
    -1,
  );

  const curried_types_declarations = keys.reduce<dts.ITypeDeclaration[][]>(
    (reduced, key) => [
      ...reduced,
      create_curried_types(`${name}_${key}`, types[key], false, placeholder),
    ],
    [],
  );

  const masks = create_masks(parameters_length);

  const types_parameters = keys.reduce<dts.IParameterDeclaration[][]>(
    (reduced, key) => [...reduced, types[key].parameters!],
    [],
  );

  const non_conflicts = [...new Array(parameters_length)].map((_, index) => {
    const parameter = types_parameters[0][index];
    return types_parameters
      .slice(1)
      .every(type_parameters => R.equals(type_parameters[index], parameter));
  });

  const conflict_declarations: dts.ITypeDeclaration[][] = [];
  const conflict_declarations_selectables: dts.IObjectMember[][][] = [];

  const non_conflict_rebases: { [name: string]: string } = {};
  const non_conflict_declarations: dts.ITypeDeclaration[] = [];
  const non_conflict_declarations_selectables: dts.IObjectMember[][] = [];

  curried_types_declarations[0].forEach((curried_type_declaration, index) => {
    const is_return_signature =
      index === curried_types_declarations[0].length - 1;
    const mask = masks[index];
    if (
      !is_return_signature &&
      mask
        .split('')
        .every((bit, bit_index) => bit === '0' || non_conflicts[bit_index])
    ) {
      const rebase_name = `${name}_${mask}`;
      non_conflict_declarations.push(
        dts.create_type_declaration({
          name: rebase_name,
          generics: curried_type_declaration.generics,
          type: dts.create_object_type({
            members: curried_types_declarations
              .map(
                curried_types_declaration =>
                  (curried_types_declaration[index].type! as dts.IObjectType)
                    .members! as dts.IObjectMember[],
              )
              .reduce((a, b) => [...a, ...b], []),
          }),
        }),
      );
      non_conflict_declarations_selectables.push(
        curried_types_declarations
          .map((curried_types_declaration, type_index) =>
            create_selectable_signatures(
              (curried_types_declaration[index].type! as dts.IObjectType)
                .members! as dts.IObjectMember[],
              keys[type_index],
            ),
          )
          .reduce((a, b) => [...a, ...b], []),
      );
      curried_types_declarations
        .map(x => x[index])
        .forEach(type_declaration => {
          non_conflict_rebases[type_declaration.name] = rebase_name;
        });
    } else {
      conflict_declarations.push(
        curried_types_declarations.map(
          curried_type_declarations => curried_type_declarations[index],
        ),
      );
      conflict_declarations_selectables.push(
        curried_types_declarations.map(curried_type_declarations => {
          const members = (curried_type_declarations[index]
            .type as dts.IObjectType).members;
          return members !== undefined && members.every(is_object_member)
            ? create_selectable_signatures(members as dts.IObjectMember[])
            : [];
        }),
      );
    }
  });

  const mixed_type_declarations_selectables = [
    ...non_conflict_declarations_selectables,
    ...conflict_declarations_selectables.reduce((a, b) => [...a, ...b], []),
  ];

  return [
    ...non_conflict_declarations,
    ...conflict_declarations.reduce((a, b) => [...a, ...b], []),
  ].map((type_declaration, index) => {
    const object_type = type_declaration.type;
    if (
      !is_object_type(object_type) ||
      object_type.members!.some(member => !is_object_member(member))
    ) {
      return type_declaration;
    }
    const repeated_function_types: { [name: string]: dts.IFunctionType } = {};
    const members: dts.IObjectMember[] = [];
    object_type.members!.forEach(
      (member: dts.IObjectMember, member_index, origin_members) => {
        const function_type = (member.owned as dts.IFunctionDeclaration).type!;
        const function_return_type = function_type.return! as dts.IGeneralType;
        if (function_return_type.name in non_conflict_rebases) {
          repeated_function_types[
            non_conflict_rebases[function_return_type.name]
          ] = function_type;
        } else {
          members.push(
            member,
            ...(selectable && origin_members.length > 1
              ? [mixed_type_declarations_selectables[index][member_index]]
              : []),
          );
        }
      },
    );
    const non_conflict_merged_members = Object.keys(
      repeated_function_types,
    ).map(key => {
      const function_type = repeated_function_types[key];
      const function_return_type = function_type.return as dts.IGeneralType;
      return dts.create_object_member({
        owned: dts.create_function_declaration({
          name: undefined,
          type: {
            ...function_type,
            return: dts.create_general_type({
              name: key,
              generics: function_return_type.generics,
            }),
          },
        }),
      });
    });
    members.push(...non_conflict_merged_members);
    if (selectable) {
      members.push(
        ...create_selectable_signatures(non_conflict_merged_members),
      );
    }
    object_type.members = members;
    return type_declaration;
  });
};
