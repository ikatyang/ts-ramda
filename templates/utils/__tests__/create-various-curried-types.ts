import * as dts from 'dts-element';
import {create_various_curried_types} from '../create-curried-various-types';

const test_cases: Record<string, string> = {
  'non-generics': `
    function number(a: number, b: number): boolean;
    function string(a: string, b: string): boolean;
  `,
  'generics + multi-parameters': `
    function string(str: string): string;
    function list<T>(list: List<T>): T | undefined;
  `,
  'generics + multi-parameters + return-generics': `
    function list<T, U>(fn: Morphism<T, U>, list: List<T>): U[];
    function functor<T, U>(fn: Morphism<T, U>, functor: Functor<T>): Functor<U>;
  `,
};

it('should throw error if those function types have different length', () => {
  expect(() => create_various_curried_types('$', parse_types(`
    function $1(a: any): any;
    function $2(a: any, b: any): any;
  `))).toThrowError();
});

Object.keys(test_cases).forEach(case_name => {

  describe(case_name, () => {

    it('should transform correctly without any options', () => {
      expect(emit_curried_types()).toMatchSnapshot();
    });

    it('should transform correctly with selectable', () => {
      expect(emit_curried_types({
        selectable: true,
      })).toMatchSnapshot();
    });

    it('should transform correctly with placeholder', () => {
      expect(emit_curried_types({
        placeholder: true,
      })).toMatchSnapshot();
    });

    it('should transform correctly with every option', () => {
      expect(emit_curried_types({
        selectable: true,
        placeholder: true,
      })).toMatchSnapshot();
    });
  });

  function emit_curried_types(options: {selectable?: boolean, placeholder?: boolean} = {}) {
    const test_case = test_cases[case_name];
    const function_types = parse_types(test_case);

    const {
      selectable = false,
      placeholder = false,
    } = options;

    const curried_types = create_various_curried_types('$', function_types, selectable, placeholder);
    const top_level_element = dts.create_top_level_element({members: curried_types});
    return dts.emit(top_level_element);
  }
});

function parse_types(types_string: string) {
  const function_declarations = dts.parse(types_string).members as dts.IFunctionDeclaration[];
  return function_declarations.reduce<Record<string, dts.IFunctionType>>(
    (current, function_declaration) => ({
      ...current,
      [function_declaration.name!]: function_declaration.type!,
    }),
    {},
  );
}
