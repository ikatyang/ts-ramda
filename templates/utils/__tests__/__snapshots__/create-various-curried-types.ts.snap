// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generics + multi-parameters + return-generics should transform correctly with every option 1`] = `
"type $_00 = {
    <T$1, U, T$2 extends T$1 = T$1>(fn: Morphism<T$1, U>, list: List<T$2>): $_list_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T$1, U, T$2 extends T$1 = T$1>(fn: Morphism<T$1, U>, list: List<T$2>) => $_list_11<U>;
    <T>(_fn: PH, list: List<T>): $_list_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"list\\">(): <T>(_fn: PH, list: List<T>) => $_list_01<T>;
    <T$1, U, T$2 extends T$1 = T$1>(fn: Morphism<T$1, U>, functor: Functor<T$2>): $_functor_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T$1, U, T$2 extends T$1 = T$1>(fn: Morphism<T$1, U>, functor: Functor<T$2>) => $_functor_11<U>;
    <T>(_fn: PH, functor: Functor<T>): $_functor_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"functor\\">(): <T>(_fn: PH, functor: Functor<T>) => $_functor_01<T>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
};
type $_10<T, U> = {
    <T$1 extends T = T>(list: List<T$1>): $_list_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): <T$1 extends T = T>(list: List<T$1>) => $_list_11<U>;
    <T$1 extends T = T>(functor: Functor<T$1>): $_functor_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): <T$1 extends T = T>(functor: Functor<T$1>) => $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly with late-inference 1`] = `
"type $_00 = {
    <T$1, U, T$2 extends T$1 = T$1>(fn: Morphism<T$1, U>, list: List<T$2>): $_list_11<U>;
    <T$1, U, T$2 extends T$1 = T$1>(fn: Morphism<T$1, U>, functor: Functor<T$2>): $_functor_11<U>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
};
type $_10<T, U> = {
    <T$1 extends T = T>(list: List<T$1>): $_list_11<U>;
    <T$1 extends T = T>(functor: Functor<T$1>): $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly with placeholder 1`] = `
"type $_00 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <T>(_fn: PH, list: List<T>): $_list_01<T>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
    <T>(_fn: PH, functor: Functor<T>): $_functor_01<T>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly with selectable 1`] = `
"type $_00 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_list_11<U>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T, U>(fn: Morphism<T, U>, functor: Functor<T>) => $_functor_11<U>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): (list: List<T>) => $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): (functor: Functor<T>) => $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly without any options 1`] = `
"type $_00 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters should transform correctly with every option 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (str: string) => $_string_1;
    <T>(list: List<T>): $_list_1<T>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): <T>(list: List<T>) => $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`generics + multi-parameters should transform correctly with late-inference 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <T>(list: List<T>): $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`generics + multi-parameters should transform correctly with placeholder 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <T>(list: List<T>): $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`generics + multi-parameters should transform correctly with selectable 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (str: string) => $_string_1;
    <T>(list: List<T>): $_list_1<T>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): <T>(list: List<T>) => $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`generics + multi-parameters should transform correctly without any options 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <T>(list: List<T>): $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`non-generics should transform correctly with every option 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    <$SEL extends \\"11\\", $KIND extends \\"number\\">(): (a: number, b: number) => $_number_11;
    (a: number): $_number_10;
    <$SEL extends \\"1\\", $KIND extends \\"number\\">(): (a: number) => $_number_10;
    (_a: PH, b: number): $_number_01;
    <$SEL extends \\"01\\", $KIND extends \\"number\\">(): (_a: PH, b: number) => $_number_01;
    (a: string, b: string): $_string_11;
    <$SEL extends \\"11\\", $KIND extends \\"string\\">(): (a: string, b: string) => $_string_11;
    (a: string): $_string_10;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (a: string) => $_string_10;
    (_a: PH, b: string): $_string_01;
    <$SEL extends \\"01\\", $KIND extends \\"string\\">(): (_a: PH, b: string) => $_string_01;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;

exports[`non-generics should transform correctly with late-inference 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    (a: number): $_number_10;
    (a: string, b: string): $_string_11;
    (a: string): $_string_10;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;

exports[`non-generics should transform correctly with placeholder 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    (a: number): $_number_10;
    (_a: PH, b: number): $_number_01;
    (a: string, b: string): $_string_11;
    (a: string): $_string_10;
    (_a: PH, b: string): $_string_01;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;

exports[`non-generics should transform correctly with selectable 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    <$SEL extends \\"11\\", $KIND extends \\"number\\">(): (a: number, b: number) => $_number_11;
    (a: number): $_number_10;
    <$SEL extends \\"1\\", $KIND extends \\"number\\">(): (a: number) => $_number_10;
    (a: string, b: string): $_string_11;
    <$SEL extends \\"11\\", $KIND extends \\"string\\">(): (a: string, b: string) => $_string_11;
    (a: string): $_string_10;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (a: string) => $_string_10;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;

exports[`non-generics should transform correctly without any options 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    (a: number): $_number_10;
    (a: string, b: string): $_string_11;
    (a: string): $_string_10;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;
